Moonscraper Chart Editor/Assets/Scripts/Game/Misc/PhraseValidator.cs

```diff
@@ -0,0 +1,55 @@
+﻿using System.Collections.Generic;
+using MoonscraperChartEditor.Song;
+using Event = MoonscraperChartEditor.Song.Event;
+
+namespace Game.Misc
+{
+    public static class PhraseValidator
+    {
+        public static List<IValidationError> ValidateIntegrity(Song song)
+        {
+            var validationErrors = new List<IValidationError>();
+            Event previousEvent = null;
+            foreach (var songEvent in song.eventsAndSections)
+            {
+                if (songEvent is Section) continue;
+
+                switch (songEvent.title)
+                {
+                    case PhraseEvent.PhraseStart: {
+                        if (PhraseEvent.PhraseStart.Equals(previousEvent?.title))
+                        {
+                            validationErrors.Add(new MismatchedPhraseEventError(previousEvent));
+                        }
+
+                        break;
+                    }
+                    case PhraseEvent.PhraseEnd: {
+                        if (!PhraseEvent.PhraseStart.Equals(previousEvent?.title))
+                        {
+                            validationErrors.Add(new MismatchedPhraseEventError(songEvent));
+                        }
+
+                        break;
+                    }
+                    default:
+                        continue;
```

Comment:
This whole loop is a bit hard to read with all the jumping around it does, so could probably do with a slight refactor or some comments but overall this does work so I'm not too fussed about it.

Reply:
I tried to reduce complexity of the method by reducing the nesting within and being more explicit about what I'm checking.